File: 0438. Find All Anagrams in a String.cpp
class Solution
{
public:
    vector<int> findAnagrams(string s, string p)
    {
        if (s.size() < p.size())
        {
            return { };
        }
        
        vector<int> pl(26, 0);
        vector<int> sl(26, 0);
        
        for (int i = 0; i < p.size(); ++i)
        {
            ++pl[p[i] - 'a'];
            ++sl[s[i] - 'a'];
        }
        
        vector<int> ans;
        
        if (pl == sl)
        {
            ans.push_back(0);
        }
        
        for (int i = p.size(); i < s.size(); ++i)
        {
            --sl[s[i - p.size()] - 'a'];
            ++sl[s[i] - 'a'];
            
            if (pl == sl)
            {
                ans.push_back(i - p.size() + 1);
            }
        }
        
        return ans;
    }
};

File: 0438. Find All Anagrams in a String.cs
public class Solution
{
    public IList<int> FindAnagrams(string s, string p)
    {
        var frequencies = new int[26];

        foreach (var c in p)
        {
            ++frequencies[c - 'a'];
        }

        var l = 0;
        var r = 0;
        var currentFrequencies = new int[26];
        var result = new List<int>();

        while (r < s.Length)
        {
            var current = s[r] - 'a';
            ++currentFrequencies[current];

            while (currentFrequencies[current] > frequencies[current])
            {
                --currentFrequencies[s[l] - 'a'];
                ++l;
            }

            if (frequencies.SequenceEqual(currentFrequencies))
            {
                result.Add(l);
            }

            ++r;
        }

        return result;
    }
}